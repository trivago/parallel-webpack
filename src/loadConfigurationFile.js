import { jsVariants } from 'interpret';
import { endsWith } from 'lodash';
import chalk from 'chalk';

export const availableExtensions = Object.keys(jsVariants).sort((a, b) => {
    // sort extensions to ensure that .babel.js and
    // similar ones are always matched before .js
    const res = -(a.split(/\./).length - b.split(/\./).length);
    // all things being equal, we need to
    // prioritize .js as it is most likely
    if (res === 0) {
        if (a === '.js') {
            return -1;
        }
        if (b === '.js') {
            return 1;
        }
        return 0;
    }
    return res;
});

export const getMatchingLoader = (
    configPath,
    extensions = availableExtensions,
    variants = jsVariants,
) => {
    const ext = extensions.find(ext => endsWith(configPath, ext));
    return ext ? variants[ext] : null;
};

const callConfigFunction = fn =>
    fn(require('minimist')(process.argv, { '--': true }).env || {});

const getConfig = configPath => {
    const configModule = require(configPath);
    const configDefault =
        configModule && configModule.__esModule
            ? configModule.default
            : configModule;
    return typeof configDefault === 'function'
        ? callConfigFunction(configDefault)
        : configDefault;
};

export default (configPath, getMatchingLoaderFn = getMatchingLoader) => {
    const mod = getMatchingLoaderFn(configPath);

    if (mod) {
        const mods = Array.isArray(mod) ? mod : [mod];
        const installed = mods.some(mod => {
            if (typeof mod === 'string') {
                try {
                    require(mod);
                    return true;
                } catch (ignored) {}
            } else if (typeof mod === 'object') {
                try {
                    var s = require(mod.module);
                    mod.register(s);
                    return true;
                } catch (ignored) {}
            }
        });

        if (!installed) {
            throw new Error(
                'Could not load required module loading for ' +
                    chalk.underline(configPath),
            );
        }
    }
    return getConfig(configPath);
};
