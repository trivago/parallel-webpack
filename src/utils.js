import Ajv from 'ajv';
import schema from '../schema.json';
import loadConfigurationFile from './loadConfigurationFile';
import { decorate } from './logger';
import chalk from 'chalk';

export const validateOptions = farmOptions => {
    const validate = new Ajv({
        allErrors: true,
        coerceTypes: true,
        removeAdditional: 'all',
        useDefaults: true,
    }).compile(schema);

    if (!validate(farmOptions)) {
        const errorMessage = validate.errors
            .map(({ dataPath, message }) => {
                return `Property: "options ${dataPath}" ${message}`;
            })
            .join('\n');
        return Promise.reject(
            `Options validation failed:
            ${errorMessage}
            `,
        );
    }
    return Promise.resolve('Options validation succeeded.');
};
export const loadConfig = (configPath, options) => {
    const argvBackup = process.argv;
    try {
        process.argv = options.argv;
        const config = loadConfigurationFile(configPath);
        process.argv = argvBackup;
        return Promise.resolve(config);
    } catch (e) {
        process.argv = argvBackup;
        return Promise.reject(
            new Error(
                decorate(
                    `Could not load configuration file ${chalk.underline(
                        configPath,
                    )} \n ${e}`,
                    'red',
                ),
            ),
        );
    }
};
