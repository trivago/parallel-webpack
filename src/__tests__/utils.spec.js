import { loadConfig, validateOptions } from '../utils';
import loadConfigurationFile from '../loadConfigurationFile';
jest.mock('../loadConfigurationFile');

describe('Utility functions', () => {
    describe('loadConfig', () => {
        const testConfig = {
            name: 'testconfig',
        };
        beforeEach(() => {
            jest.clearAllMocks();
        });
        it('should resolve when load is successful', () => {
            loadConfigurationFile.mockImplementationOnce(() => testConfig);
            expect(loadConfig('/path/to/webpack', {})).resolves.toEqual(
                testConfig,
            );
        });
        it('should reject when load is unsuccessful', () => {
            loadConfigurationFile.mockImplementationOnce(() => {
                throw new Error('File not existing');
            });
            expect(
                loadConfig('/path/to/webpack', {}),
            ).rejects.toMatchSnapshot();
        });
    });

    describe('validateOptions', () => {
        it('should resolve when validation is successful', () => {
            expect(
                validateOptions({
                    maxRetries: 3,
                    numWorkers: 6,
                    forkOptions: {
                        stdio: 'pipe',
                    },
                }),
            ).resolves.toBe('Options validation succeeded.');
        });

        it('should reject with error when validation is unsuccessful', () => {
            expect(
                validateOptions({
                    maxRetries: '3',
                    forkOptions: [],
                }),
            ).rejects.toMatchSnapshot();
        });
    });
});
