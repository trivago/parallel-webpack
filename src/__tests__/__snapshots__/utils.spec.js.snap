// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utility functions loadConfig should reject when load is unsuccessful 1`] = `
[Error: [31m[WEBPACK][39m Could not load configuration file [4m/path/to/webpack[24m 
 Error: File not existing]
`;

exports[`Utility functions validateOptions should reject with error when validation is unsuccessful 1`] = `
"Options validation failed:
            Property: \\"options .forkOptions\\" should be object
            "
`;
